File: 202311/2147.go
func numberOfWays(corridor string) int {
	n := len(corridor)
	suf := make([]int, n)
	if corridor[n-1] == 'S' {
		suf[n-1] = 1
	}
	for i := n - 2; i >= 0; i-- {
		if corridor[i] == 'S' {
			suf[i] = suf[i+1] + 1
		} else {
			suf[i] = suf[i+1]
		}
	}
	if suf[0]%2 == 1 {
		return 0
	}
	const MOD = int(1e9 + 7)
	dp := make([]int, n)
	for i := 0; i < n; i++ {
		dp[i] = -1
	}

	var f func(k int) int
	f = func(k int) int {
		if k == n || suf[k] == 2 {
			return 1
		}
		if dp[k] != -1 {
			return dp[k]
		}
		cnt := 0
		j := -1
		for i := k; i < n; i++ {
			if corridor[i] == 'S' {
				cnt++
			}
			if cnt == 2 {
				j = i
				break
			}
		}
		if j == -1 {
			dp[k] = 0
			return dp[k]
		}
		ans := f(j + 1)
		for i := j + 1; i < n; i++ {
			if corridor[i] == 'P' {
				ans = (ans + f(i+1)) % MOD
			} else {
				break
			}
		}
		dp[k] = ans
		return ans
	}

	return f(0)
}

File: 202512/2147.py
class Solution:
    def numberOfWays(self, corridor: str) -> int:
        n = len(corridor)
        MOD = 10**9 + 7
        cnt = corridor.count("S")
        if cnt % 2 == 1 or cnt < 2:
            return 0
        if cnt == 2:
            return 1

        ans = 1
        done = 0
        i = 0
        j, k = -1, -1
        while i < n:
            if corridor[i] == "S":
                done += 1
                if done == 2:
                    j = i
                if done == 3:
                    k = i
                    done = 1
                    ans = ans * (k - j) % MOD
            i += 1
        return ans
